### PR TITLE
enc264: implement I/P/B QP setting on CQP mode

### PR DESCRIPTION
**Pull request by wudping**
set slice_qp_delta acording to the slice type.

Signed-off-by: wudping dongpingx.wu@intel.com

### CODE REVIEW

**Comment by chivakker** [comment]
lgtm :+1:

**Comment by xuguangxin** [comment]
Hi @lizhong1008 
Please help to review this.One question from me
1. Do we need change diffIPQP  to more descriptive name? Do we need refer to x264’s name?

File: encoder/vaapiencoder_base.cpp [hunk 1]
```diff
@@ -57,6 +57,8 @@ VaapiEncoderBase::VaapiEncoderBase():
     m_videoParamCommon.rcParams.windowSize = 500;
     m_videoParamCommon.rcParams.disableBitsStuffing = 1;
     m_videoParamCommon.leastInputCount = 0;
+    m_videoParamCommon.rcParams.diffIPQP = 2;
```
Comment (xuguangxin) [inline encoder/vaapiencoder_base.cpp]: What is the 2 and 4 come from? Will it introduce quality problem to default encoder param?

**Comment by lizhong1008** [comment]
@xuguangxin seem x264 won't provide direct interface to set P/B frame qp, it use --qpfile to set different frame_type qp. I think it ok to use the name of "diffIPQP" if well commented.

File: encoder/vaapiencoder_base.cpp (hunk 1 above)
Comment (lizhong1008) [inline encoder/vaapiencoder_base.cpp]: Default diffIPQP and diffPBQP should be set to 0.

File: interface/VideoEncoderDefs.h [hunk 2]
```diff
@@ -189,10 +189,14 @@ typedef struct VideoRateControlParams {
     uint32_t targetPercentage;
     uint32_t disableFrameSkip;
     uint32_t disableBitsStuffing;
+    uint32_t diffIPQP;// I/P frame QP difference
+    uint32_t diffPBQP;// P/B frame QP difference,
```
Comment (lizhong1008) [inline interface/VideoEncoderDefs.h]: How about use the name of diffIBQP ?

File: interface/VideoEncoderDefs.h [hunk 3]
```diff
@@ -189,10 +189,14 @@ typedef struct VideoRateControlParams {
     uint32_t targetPercentage;
     uint32_t disableFrameSkip;
     uint32_t disableBitsStuffing;
+    uint32_t diffIPQP;// I/P frame QP difference
```
Comment (lizhong1008) [inline interface/VideoEncoderDefs.h]: Would better change the comment to "P frame qp  minus initQP"  to make it clearer.

File: encoder/vaapiencoder_h264.cpp [hunk 4]
```diff
@@ -1356,6 +1356,15 @@ bool VaapiEncoderH264::addSliceHeaders (const PicturePtr& picture) const
...
         sliceParam->slice_qp_delta = initQP() - minQP();
+        switch (picture->m_type) {
+        case VAAPI_PICTURE_B:
+            sliceParam->slice_qp_delta += m_videoParamCommon.rcParams.diffPBQP;
+        case VAAPI_PICTURE_P:
+            sliceParam->slice_qp_delta += m_videoParamCommon.rcParams.diffIPQP;
+        case VAAPI_PICTURE_I:
+        default:
+            break;
+        }
```
Comment (lizhong1008) [inline encoder/vaapiencoder_h264.cpp:1377]: We need to check the I/P/B frame qp is in the range of maxQP and minQp.

File: interface/VideoEncoderDefs.h (hunk 2 above)
Comment (wudping) [inline interface/VideoEncoderDefs.h]: okay, Thanks LiZhong.

File: interface/VideoEncoderDefs.h (hunk 3 above)
Comment (wudping) [inline interface/VideoEncoderDefs.h]: ok, Thanks LiZhong.

File: encoder/vaapiencoder_h264.cpp (hunk 4 above)
Comment (wudping) [inline encoder/vaapiencoder_h264.cpp:1377]: okay, Thanks LiZhong.

**Comment by xuguangxin** [comment]
Hi @lizhong1008, Is it ok for this patch?

File: interface/VideoEncoderDefs.h [hunk 5]
```diff
@@ -189,10 +189,14 @@ typedef struct VideoRateControlParams {
     uint32_t targetPercentage;
     uint32_t disableFrameSkip;
     uint32_t disableBitsStuffing;
+    uint32_t diffQPIP;// P frame qp minus initQP
+    uint32_t diffQPIB;// B frame qp minus initQP
```
Comment (lizhong1008) [inline interface/VideoEncoderDefs.h]: why unsigned int? they can be negative values.

Comment (wudping) [inline interface/VideoEncoderDefs.h]: okay, thanks LiZhong.

**Comment by lizhong1008** [comment]
+1

**Comment by xuguangxin** [comment]
@wudping , we need bump version for this

**Comment by xuguangxin** [comment]
@wudping , never mind, i will handle the version bump

**Comment by xuguangxin** [comment]
merged as 3147d364aa30b405cc85554be3f629d223ececf0

**Pull request closed without merging**